servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SingleThreadedExecutor.java

```diff
@@ -0,0 +1,202 @@
+                    }
+                 } else {
+                    thread.queue.put(command);
+                }
+            } catch (InterruptedException woken) {
+                // Interrupt status is not cleared
+                throw new RejectedExecutionException(thread.getName() + ": Task enqueue was interrupted", woken);
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        if (thread.isAlive() &&
+                WorkerThread.stateUpdater.compareAndSet(thread, WorkerThread.OPEN, WorkerThread.CLOSED) &&
+                interruptOnClose &&
+                Thread.currentThread() != thread) {
+            thread.interrupt();
+        }
+    }
+
+    private static class WorkerThread extends AsyncContextHolderThread {
+        static final AtomicIntegerFieldUpdater<WorkerThread> stateUpdater =
+                newUpdater(WorkerThread.class, "state");
+        static final int OPEN = 0; // running tasks
+        static final int CLOSED = 1; // executor has closed
+        static final int ORPHANED = 2; // executor GCed without close()
+        static final int EXITED = 3; // Worker thread exit
+
+        final BlockingQueue<Runnable> queue;
+        /**
+         * If the executor is discarded and GCed this reference will be cleared and the worker thread can exit.
+         */
+        private final WeakReference<SingleThreadedExecutor> owner;
+
+        volatile int state;
+
+        WorkerThread(SingleThreadedExecutor owner, Supplier<BlockingQueue> queueSupplier, String name) {
+            super(name);
+            this.queue = queueSupplier.get();
+            this.owner = new WeakReference<>(owner);
+
+            if (!isDaemon()) {
+                setDaemon(true);
+            }
+
+            if (Thread.NORM_PRIORITY != getPriority()) {
+                setPriority(Thread.NORM_PRIORITY);
+            }
+        }
+
+        public void run() {
+            while (OPEN == state && null != owner.get()) {
+                try {
+                    // poll vs. take because we need to occasionally check if closed or abandoned
+                    Runnable command = queue.poll(1L, TimeUnit.MINUTES);
+                    if (null == command) {
+                        continue;
+                    }
+                    try {
+                        command.run();
+                    } catch (Throwable all) {
+                        LOGGER.warn("Uncaught throwable from command {}", command, all);
+                    }
+                } catch (InterruptedException woken) {
+                    Thread.interrupted();
+                    // closing?
+                } catch (Throwable all) {
+                    LOGGER.warn("Uncaught throwable in worker thread", all);
+                }
+            }
+            int was = stateUpdater.getAndSet(this, null != owner.get() ? ORPHANED : EXITED);
+            LOGGER.debug("WorkerThread exiting " + (CLOSED == was ? "cleanly" : "messily"));
```

Comment:
Probably the log should be a warning if the exit was not clean. The extra states are indeed for debuggability, so that a conditional breakpoint can be triggered on the value being set or being present at a particular point.